joomla/joomla-cms
# Images loading=lazy, part 2n

This PR solves a lot of bugs and brings feature parity to J3, keep reading

### Summary of Changes
#### Bug fixes:
- Changes the a-tag to button in the media modals
- Restores width and height params in the media manager object (in 2 places)
- Separates the image selection script from the media script (if you had an Utd-button without a media field the button was foobar)

#### New features (parity with J3, basically missing in J4):
- Ability to enter ALT text for any image (used in the editors)
- Ability to set if an image will be lazy loaded in the content area (eg inside an editor)
- Ability to set ALT text for images inserted through drag and drop in tinyMCE
- Ability to set if an image will be lazy loaded for images inserted through drag and drop in tinyMCE
- New helper `HTMLHelper::cleanImageURL`, explainer below

#### Lazyloading
The concept is simple: it just works!
- For any image inserted in an editor (tinyMCE, Codemirror, none, or any 3rd PT) the user **has the ability to select if the image will be lazy loaded**
- For images used though the media field there's a helper to... wait for it, help you. By default the URL that is stored in the database has 2 extra URL parameters `joomla_image_width` and `joomla_image_height` that represent the width and height of the image. Passing the URL in the helper you'll get back an object with the clean url (without the 2 extra params mentioned before, and the attributes (array of width and height). So since we're already in an overridable file (layout) anyone can decide if the position of the image should or not be lazy loaded. Reading the fulltext layout is more explanatory than reading some text out of context: https://github.com/joomla/joomla-cms/blob/3427a9ca0bb265f72cb8da9a61c508a0a5859afa/layouts/joomla/content/full_image.php


### Testing Instructions
Download the package from the bottom part of the PR in Github or do it through git command

Test the tinyMCE dnd
Make sure tinyMCE is your editor
Try to insert an image in an article with dnd, enter some alt text and leave the lazy loaded checked
Repeat but unselect the lasyloaded and leave the alt empty
Toggle the editor and check the produced HTML

Test the tinyMCE media button
Try to insert an image in an article with the media button, enter some alt text and leave the lazy loaded checked
Repeat but unselect the lasyloaded and leave the alt empty
Toggle the editor and check the produced HTML

Repeat the same for Codemirror and none

Custom Fields
setup a media custom field
setup a imagelist custom field
make sure the custom fields work

check the html, should have attributes: loading=lazy, width, height for all the images.

Media field
Insert an full image
Check the html, should have attributes: loading=lazy, width, height

Insert an intro image
Check the html, should have attributes: loading=lazy, width, height


### Preview
XTD-Button:
![Screenshot 2020-09-26 at 22 27 36](https://user-images.githubusercontent.com/3889375/94349730-a1abc900-0047-11eb-92f5-22db8786eb8d.png)


TinyMCE drag and drop
![Screenshot 2020-09-26 at 22 28 12](https://user-images.githubusercontent.com/3889375/94349734-ab353100-0047-11eb-9733-d470c6c0029e.png)



### Documentation Changes Required
Document how the media field url should be handled.
Provide some guidelines for styling the lazy loaded images, basically explain how to use the css property `object-fit: scale-down, cover, fill, etc`  https://caniuse.com/object-fit to get the same responsive behaviour as without defined width,height

A good resource for width/height from Jen Simmons @jensimmons : https://speakerdeck.com/jensimmons/using-the-html-width-and-height-attributes-to-improve-web-page-loading

#### For devs and implementors:

Images selected with the media field will have by default some url params eg: `images/joomla_black.png?joomla_image_width=225&joomla_image_height=50`. Leaving the URL like that in your layouts is totally fine, Joomla is exercising the same technic for stylesheets and script to invalidate the cache (eg: `src="/media/system/js/core.min.js?b276ec42f4135ff7408390a982e26604"`). That said if you decide to skip couple of lines of code you're missing some significant performance gains from lazily loading the images. All it takes is:

- Pass the URL to the helper:
```php
$img = HTMLHelper::cleanImageURL($images->image_fulltext);

/**
 * $img->url // The clean URL, no extra params
 * $img->attributes // Array
 * $img->attributes['width']  // The width of the image
 * $img->attributes['height']  // The height of the image
 **/
```

- Set the lazy loading attribute:
```php
$img->attributes['loading'] = 'lazy';
```

- Render the image:
```php
<img src="<?php echo htmlspecialchars($img->url, ENT_COMPAT, 'UTF-8'); ?>" alt="" <?php echo ArrayHelper::toString($img->attributes); ?> />
```
